addons/hr/models/hr_employee.py

```diff
@@ -134,6 +134,56 @@ def _default_image(self):
         ('user_uniq', 'unique (user_id, company_id)', "A user cannot be linked to multiple employees in the same company.")
     ]
 
+    @api.multi
+    def name_get(self):
+        try:
+            return super(HrEmployeePrivate, self).name_get()
+        except AccessError:
+            return self.env['hr.employee.public'].browse(self.ids).name_get()
+
+    @api.multi
+    def read(self, fields, load='_classic_read'):
+        if self.check_access_rights('read', raise_exception=False):
+            return super(HrEmployeePrivate, self).read(fields, load=load)
+        else:
+            public_fields = set(self.env['hr.employee.public']._fields.keys()).intersection(fields)
+            if not public_fields:
+                raise AccessError(_('All the fields you try to access are private but you cannot read private fields on employees'))
+            # We fill all the fields with False as a value in order to stay coherent with what was asked
+            base = dict.fromkeys(fields, False)
+            return [{**base, **result} for result in self.env['hr.employee.public'].browse(self.ids).read(public_fields, load=load)]
+
+    @api.model
+    def _search(self, args, offset=0, limit=None, order=None, count=False, access_rights_uid=None):
+        """ We override the _search because it is the method that checks the access rights """
+        if self.check_access_rights('read', raise_exception=False):
+            return super(HrEmployeePrivate, self)._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
+        else:
+            return self.env['hr.employee.public']._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
+
+    @api.multi
+    def get_formview_id(self, access_uid=None):
+        """ Override this method in order to redirect many2one towards the right model depending on access_uid """
+        if access_uid:
+            self = self.sudo(access_uid)
```

Comment:
don't replace `self` otherwise you lose the original sudo() when calling get_formview_id() again below